smarts/sstudio/genhistories.py

```diff
@@ -277,34 +293,96 @@ def column_val_in_row(self, row, col_name: str):
                 dx = float(self._next_row["x"]) - float(row["x"])
                 dy = float(self._next_row["y"]) - float(row["y"])
                 dm = np.linalg.norm((dx, dy))
-                if dm > 0.0:
+                if dm != 0.0 and (
+                    self._heading_min_speed is None or dm > self._heading_min_speed
+                ):
                     r = math.atan2(dy / dm, dx / dm)
-                    return (r - math.pi / 2) % (2 * math.pi)
+                    new_heading = (r - math.pi / 2) % (2 * math.pi)
+                    if self._max_angular_velocity:
+                        # XXX: could try to divide by sim_time delta here instead of assuming .1s
+                        angular_velocity = (new_heading - self._prev_heading) / 0.1
+                        if abs(angular_velocity) > self._max_angular_velocity:
+                            new_heading = (
+                                self._prev_heading
+                                + np.sign(angular_velocity)
+                                * self._max_angular_velocity
+                                * 0.1
+                            )
+                    self._prev_heading = new_heading
+                    return new_heading
             # Note: pedestrian track files won't have this
-            return float(row.get("psi_rad", 0.0)) - math.pi / 2
+            self._prev_heading = float(row.get("psi_rad", 0.0)) - math.pi / 2
+            return self._prev_heading
         # XXX: should probably check for and handle x_offset_px here too like in NGSIM
         return None
 
 
 class NGSIM(_TrajectoryDataset):
-    def __init__(self, dataset_spec, output):
+    def __init__(self, dataset_spec: Dict[str, Any], output: str):
         super().__init__(dataset_spec, output)
-        self._prev_heading = -math.pi / 2
-
-    def _cal_heading(self, window):
-        c = window[1, :2]
-        n = window[2, :2]
-        if any(np.isnan(c)) or any(np.isnan(n)):
-            return self._prev_heading
-        s = np.linalg.norm(n - c)
-        if s == 0.0:
-            return self._prev_heading
-        vhat = (n - c) / s
-        r = math.atan2(vhat[1], vhat[0])
-        self._prev_heading = (r - math.pi / 2) % (2 * math.pi)
-        return self._prev_heading
+        self._prev_heading = 3 * math.pi / 2
+        self._max_angular_velocity = dataset_spec.get("max_angular_velocity", None)
+        self._heading_window = dataset_spec.get("heading_inference_window", 2)
+        # .22 corresponds to roughly 5mph.
+        self._heading_min_speed = dataset_spec.get("heading_inference_min_speed", 0.22)
```

Comment:
Where these default values computed through Trial and Error?

Reply:
Yes.  

fyi, in the old version from last spring, I'd just hardcoded this (.22) as a constant, but now I want to allow them to set it because it will depend on the dataset noise.